packages/filesystem/src/browser/location/location-renderer.tsx

```diff
@@ -18,34 +18,87 @@ import URI from '@theia/core/lib/common/uri';
 import { LocationService } from './location-service';
 import { ReactRenderer } from '@theia/core/lib/browser/widgets/react-renderer';
 import * as React from 'react';
-
+import * as ReactDOM from 'react-dom';
+import { FileService } from '../file-service';
 export class LocationListRenderer extends ReactRenderer {
 
     protected _drives: URI[] | undefined;
+    protected doShowTextInput = false;
+    protected lastUniqueTextInputLocation: URI | undefined;
+    protected previousAutocompleteMatch: string;
+    protected doAttemptAutocomplete = true;
 
     constructor(
         protected readonly service: LocationService,
-        host?: HTMLElement
+        host?: HTMLElement,
+        protected readonly fileService?: FileService
     ) {
         super(host);
         this.doLoadDrives();
     }
 
     render(): void {
-        super.render();
+        ReactDOM.render(<React.Fragment>{this.doRender()}</React.Fragment>, this.host, this.doAfterRender);
+    }
+
+    protected doAfterRender = (): void => {
         const locationList = this.locationList;
+        const locationListTextInput = this.locationTextInput;
         if (locationList) {
             const currentLocation = this.service.location;
             locationList.value = currentLocation ? currentLocation.toString() : '';
+        } else if (locationListTextInput) {
+            setTimeout(() => locationListTextInput.focus());
         }
-    }
+    };
 
     protected readonly handleLocationChanged = (e: React.ChangeEvent<HTMLSelectElement>) => this.onLocationChanged(e);
+    protected readonly handleLocationKeydown = (e: React.KeyboardEvent<HTMLSelectElement>) => this.onLocationKeydown(e);
+    protected readonly handleTextInputOnChange = (e: React.ChangeEvent<HTMLInputElement>) => this.onTextInputChanged(e);
+    protected readonly handleTextInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => this.onTextInputKeyDown(e);
+    protected readonly handleTextInputOnBlur = () => this.onTextInputToggle(false);
+    protected readonly handleTextInputMouseDown = (e: React.MouseEvent<HTMLSpanElement>) => this.onTextInputToggle(e.currentTarget.id === 'select-input');
+
     protected doRender(): React.ReactNode {
```

Comment:
There's a fair bit going on in this method now, so it might benefit from some refactoring. Maybe something like:
```typescript
<>
{!!this.fileService && this.getInputIcons()} // or with the conditional inside the the method
{this.doShowTextInput
  ? this.getTextInput()
  : this.getDropdown()
}
```
That way the high level logic of the conditionals is clear at a glance, and then each subcomponent can be inspected in detail.

Reply:
Sounds good, I've made that change